### ISSUE TITLE
glooctl json output for grpc services is not working since 1.14.0

### ISSUE THREAD

**Issue by day0ops** [solo-io/gloo#9743]
### Gloo Edge Product

Enterprise

### Gloo Edge Version

1.16.x

### Kubernetes Version

1.26.x

### Describe the bug

Since the API changes in `1.14.0-beta17` (as per https://github.com/solo-io/gloo/pull/7812) `Upstream` objects no longer holds the information about gRPC services and methods. This unfortunately causes `glooctl get upstream -o json` to filter out the gRPC services and methods.

We partially addressed this in https://github.com/solo-io/gloo/issues/9644 to fix the table format but we still need to implement a proper fix for json format as well.

### Expected Behavior

`glooctl get upstream -o json` should be able to display the gRPC services and relevant methods.

### Steps to reproduce the bug

1. Enable FDS and UDS for a grpc endpoint.
2. Follow the [guide](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/grpc/simple-routing/) on adding gRPC services.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1206768562311555/1207800332186040) by [Unito](https://www.unito.io)

**Comment by day0ops** [comment]
After speaking with @davidjumani raised this ticket to implement a proper fix.

**Comment by DuncanDoyle** [comment]
For completeness, the JSON output when running on GE 1.13.39:

```
glooctl get upstream default-grpcstore-demo-80 -o json | jq .
----------
glooctl binary version (1.16.17) differs from server components (v1.13.39) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.13.39
----------

[
  {
    "namespacedStatuses": {
      "statuses": {
        "gloo-system": {
          "state": "Accepted",
          "reportedBy": "gloo"
        }
      }
    },
    "metadata": {
      "name": "default-grpcstore-demo-80",
      "namespace": "gloo-system",
      "resourceVersion": "1739",
      "labels": {
        "discovered_by": "kubernetesplugin",
        "discovery.solo.io/function_discovery": "enabled"
      },
      "annotations": {
        "gloo.solo.io/h2_service": "true",
        "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{\"gloo.solo.io/h2_service\":\"true\"},\"labels\":{\"app\":\"grpcstore-demo\"},\"name\":\"grpcstore-demo\",\"namespace\":\"default\"},\"spec\":{\"internalTrafficPolicy\":\"Cluster\",\"ipFamilies\":[\"IPv4\"],\"ipFamilyPolicy\":\"SingleStack\",\"ports\":[{\"port\":80,\"protocol\":\"TCP\",\"targetPort\":8080}],\"selector\":{\"app\":\"grpcstore-demo\"},\"sessionAffinity\":\"None\",\"type\":\"ClusterIP\"}}\n"
      },
      "generation": "4",
      "ownerReferences": []
    },
    "discoveryMetadata": {
      "labels": {
        "app": "grpcstore-demo"
      }
    },
    "kube": {
      "serviceName": "grpcstore-demo",
      "serviceNamespace": "default",
      "servicePort": 80,
      "selector": {
        "app": "grpcstore-demo"
      },
      "serviceSpec": {
        "grpc": {
          "descriptors": "Q3F...NPQ==", 
          "grpcServices": [
            {
              "packageName": "solo.examples.v1",
              "serviceName": "StoreService",
              "functionNames": [
                "CreateItem",
                "ListItems",
                "DeleteItem",
                "GetItem"
              ]
            }
          ]
        }
      }
    },
    "useHttp2": {
      "value": true
    }
  }
]
```

And the YAML output:
```
glooctl get upstream default-grpcstore-demo-80 -o yaml
----------
glooctl binary version (1.16.17) differs from server components (v1.13.39) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.13.39
----------

---
discoveryMetadata:
  labels:
    app: grpcstore-demo
kube:
  selector:
    app: grpcstore-demo
  serviceName: grpcstore-demo
  serviceNamespace: default
  servicePort: 80
  serviceSpec:
    grpc:
      descriptors: Q3F...NPQ==  # truncated
      grpcServices:
      - functionNames:
        - CreateItem
        - ListItems
        - DeleteItem
        - GetItem
        packageName: solo.examples.v1
        serviceName: StoreService
metadata:
  annotations:
    gloo.solo.io/h2_service: "true"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"gloo.solo.io/h2_service":"true"},"labels":{"app":"grpcstore-demo"},"name":"grpcstore-demo","namespace":"default"},"spec":{"internalTrafficPolicy":"Cluster","ipFamilies":["IPv4"],"ipFamilyPolicy":"SingleStack","ports":[{"port":80,"protocol":"TCP","targetPort":8080}],"selector":{"app":"grpcstore-demo"},"sessionAffinity":"None","type":"ClusterIP"}}
  generation: "4"
  labels:
    discovered_by: kubernetesplugin
    discovery.solo.io/function_discovery: enabled
  name: default-grpcstore-demo-80
  namespace: gloo-system
  ownerReferences: []
  resourceVersion: "1739"
namespacedStatuses:
  statuses:
    gloo-system:
      reportedBy: gloo
      state: Accepted
useHttp2:
  value: true
```

**Comment by bewebi** [comment]
A "fix" for this has been merged in 1.18 and all LTS branches and will be (or has been) released in OSS 1.18.0-beta10, 1.17.1, 1.16.19 (not yet released), 1.15.30, and 1.14.32 (not yet released)

Users will need to download the latest version of glooctl for the minor version of Gloo OSS or EE that they have installed

The function names will be appended to the namespaced status in the following format for both JSON and YAML output:
```
$ glooctl get upstream default-grpcstore-demo-80 -o yaml

---
discoveryMetadata:
  labels:
    app: grpcstore-demo
kube:
  selector:
    app: grpcstore-demo
  serviceName: grpcstore-demo
  serviceNamespace: default
  servicePort: 80
  serviceSpec:
    grpcJsonTranscoder:
      matchIncomingRequestRoute: true
      protoDescriptorBin: Cqw...zM=
      services:
      - solo.examples.v1.StoreService
metadata:
...
namespacedStatuses:
  statuses:
    gloo-system:
      details:
        fields:
          functionNames:
            structValue:
              fields:
                solo.examples.v1.StoreService:
                  listValue:
                    values:
                    - stringValue: CreateItem
                    - stringValue: ListItems
                    - stringValue: DeleteItem
                    - stringValue: GetItem
      reportedBy: gloo
      state: Accepted
```

Users may want to use jq to wrangle this output to a more concise/human friendly format
Here is an example of how this can be done:
```
$ glooctl get upstream default-grpcstore-demo-80 -o json | jq '.[].namespacedStatuses.statuses.[].details.fields.functionNames.structValue.fields | to_entries | map({(.key): [.value.listValue.values.[].stringValue]})'

[
  {
    "solo.examples.v1.StoreService": [
      "CreateItem",
      "ListItems",
      "DeleteItem",
      "GetItem"
    ]
  }
]
```

**Comment by gektor0856** [comment]
Hello, my components versions:

Client: {"version":"1.15.30"}
Server: {"type":"Gateway","kubernetes":{"containers":[{"Tag":"1.15.30","Name":"discovery","Registry":"quay.io/solo-io"},{"Tag":"1.15.30","Name":"gloo-envoy-wrapper","Registry":"quay.io/solo-io"},{"Tag":"1.15.30","Name":"gloo","Registry":"quay.io/solo-io","OssTag":"1.15.30"}],"namespace":"bai-infra"}}

`glooctl get us -nbai-infra` returns that

```
unable to create proto registry files
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x454a7e5]

goroutine 1 [running]:
github.com/solo-io/gloo/projects/gloo/cli/pkg/printers.getFunctionsFromDescriptorBin(0xc000fce180)
	/workspace/gloo/projects/gloo/cli/pkg/printers/upstream.go:371 +0x125
github.com/solo-io/gloo/projects/gloo/cli/pkg/printers.addFunctionsFromGrpcTranscoder.func1(0xc000f93200)
	/workspace/gloo/projects/gloo/cli/pkg/printers/upstream.go:331 +0x131
github.com/solo-io/gloo/projects/gloo/pkg/api/v1.UpstreamList.Each({0xc000b76280, 0x26, 0x28}, 0xc000b605e8)
	/workspace/gloo/projects/gloo/pkg/api/v1/upstream.sk.go:129 +0x74
github.com/solo-io/gloo/projects/gloo/cli/pkg/printers.PrintUpstreams({0xc000b76280, 0x26, 0x28}, 0x0, 0x0, {0x20cbe9a51, 0x3})
	/workspace/gloo/projects/gloo/cli/pkg/printers/upstream.go:37 +0x14c
github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/get.Upstream.func1(0xc00092b500, {0xc000eadd20, 0x0, 0x1})
	/workspace/gloo/projects/gloo/cli/pkg/cmd/get/upstream.go:30 +0x354
github.com/spf13/cobra.(*Command).execute(0xc00092b500, {0xc000eadcf0, 0x1, 0x1})
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0xcce
github.com/spf13/cobra.(*Command).ExecuteC(0xc00092a300)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x9ed
github.com/spf13/cobra.(*Command).Execute(0xc00092a300)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992 +0x32
main.main()
	/workspace/gloo/projects/gloo/cli/cmd/main.go:11 +0x1d
```

I using proto with custom annotations such as "openapiv3/annotations.proto", "google/api/annotations.proto", "validate/validate.proto"
